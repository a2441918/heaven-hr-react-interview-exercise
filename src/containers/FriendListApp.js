import React, {Component} from 'react';
import PropTypes from 'prop-types';
import styles from './FriendListApp.css';
import {connect} from 'react-redux';

import {addFriend, deleteFriend, starFriend} from '../actions/FriendsActions';
import {FriendList, AddFriendInput, PaginationContainer, ItemsPerPage} from '../components';

class FriendListApp extends Component {

    constructor(props) {
        super(props);
        this.state = {
            pageOfItems: []
        };
    }

    onChangePage = (pageOfItems) => {
        this.setState({pageOfItems: pageOfItems});
    };

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevProps.pagingOption !== this.props.pagingOption) {
            this.renderPagination(this.props.pagingOption);
        }
    }

    renderPagination = (pagingOption) => {
        return (
            <PaginationContainer items={this.props.friendsById} onChangePage={this.onChangePage}
                                 numPerPage={pagingOption ? pagingOption : this.props.pagingOption}/>
        );
    };

    render() {
        const {friendsById} = this.props;

        const actions = {
            addFriend: this.props.addFriend,
            deleteFriend: this.props.deleteFriend,
            starFriend: this.props.starFriend
        };

        return (
            <div className={styles.friendListApp}>
                <h1>The FriendList</h1>
                <AddFriendInput addFriend={actions.addFriend}/>
                <FriendList friends={friendsById.length > 2 ? this.state.pageOfItems : friendsById} actions={actions}/>
                <div className={'paginationContainer'}>
                    <ItemsPerPage/>
                    {friendsById.length > 2 &&
                    this.renderPagination()}
                </div>
            </div>
        );
    }
}

FriendListApp.propTypes = {
    pagingOption: PropTypes.number.isRequired,
    friendsById: PropTypes.arrayOf(PropTypes.shapeOf({
        name: PropTypes.string.isRequired,
        gender: PropTypes.string.isRequired,
        starred: PropTypes.bool.isRequired,
        id: PropTypes.number
    })),
    addFriend: PropTypes.func.isRequired,
    deleteFriend: PropTypes.func.isRequired,
    starFriend: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
    return {
        pagingOption: parseInt(state.friendlist.changePagingReducer.num, 10),
        friendsById: state.friendlist.friendsReducer.friendsById
    };
};

export default connect(mapStateToProps, {
    addFriend,
    deleteFriend,
    starFriend
})(FriendListApp);
