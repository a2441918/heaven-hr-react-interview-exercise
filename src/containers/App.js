import React, {Component} from 'react';
import {combineReducers, createStore, compose } from 'redux';
import {Provider} from 'react-redux';

import FriendListApp from './FriendListApp';
import * as reducers from '../reducers';

const reducer = combineReducers(reducers);

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(reducer, {}, composeEnhancers());

export default class App extends Component {
    render() {
        return (
            <div>
                <Provider store={store}>
                    <FriendListApp/>
                </Provider>
            </div>
        );
    }
}
