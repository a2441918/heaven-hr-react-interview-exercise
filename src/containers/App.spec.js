import React from 'react';
import {shallow} from 'enzyme';

import App from './App';
import FriendListApp from './FriendListApp';

describe('App', () => {
    it('should render correctly in "debug" mode', () => {
        const wrapper = shallow(<App debug/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render FriendListApp component', () => {
        const wrapper = shallow(<App />);
        expect(wrapper.find(FriendListApp)).toHaveLength(1)
    })
});
