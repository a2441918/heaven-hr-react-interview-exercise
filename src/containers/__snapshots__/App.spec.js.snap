exports[`App should render correctly in "debug" mode 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    debug={true} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Provider
        store={
                Object {
                        "dispatch": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                        Symbol(observable): [Function],
                      }
        }>
        <Connect(FriendListApp) />
</Provider>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Connect(FriendListApp) />,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Provider
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }>
          <Connect(FriendListApp) />
</Provider>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Connect(FriendListApp) />,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
          "getChildContext": Object {
            "calledByRenderer": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
